classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * @file classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
+ *
+ * Copyright (c) 2014-2020 Simon Fraser University
+ * Copyright (c) 2000-2020 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class RepresentationUploadAccessPolicy
+ * @ingroup security_authorization_internal
+ *
+ * @brief Policy that checks whether a file can be uploaded to a representation.
+ *   It checks whether the user is allowed to access the representation file stage,
+ *   whether the representation exists, whether it matches the authorized submission,
+ *   and whether it is not part of a published publication. This policy expects an
+ *   authorized submission in the authorization context.
+ */
+
+import('lib.pkp.classes.security.authorization.DataObjectRequiredPolicy');
+
+class RepresentationUploadAccessPolicy extends DataObjectRequiredPolicy {
+
+	/** @var int */
+	public $_representationId;
+
+	/**
+	 * Constructor
+	 * @param $request PKPRequest
+	 * @param $args array request parameters
+	 * @param $representationId int
+	 */
+	function __construct($request, &$args, $representationId) {
+		parent::__construct($request, $args, 'user.authorization.accessDenied');
+		$this->_representationId = $representationId;
+	}
+
+	//
+	// Implement template methods from AuthorizationPolicy
+	//
+	/**
+	 * @see DataObjectRequiredPolicy::dataObjectEffect()
+	 */
+	function dataObjectEffect() {
+		AppLocale::requireComponents([LOCALE_COMPONENT_PKP_SUBMISSION, LOCALE_COMPONENT_APP_SUBMISSION]);
+
+		$assignedFileStageIds = $this->getAuthorizedContextObject(ASSOC_TYPE_ACCESSIBLE_FILE_STAGES);
+		if (empty($assignedFileStageIds) || !in_array(SUBMISSION_FILE_PROOF, $assignedFileStageIds)) {
+			return AUTHORIZATION_DENY;
+		}
+
+		if (empty($this->_representationId))  {
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
```

Comment:
Are references to `$msg` left-over debugging code?

Reply:
This avoids a pass-by-reference notice or something like that. I can't remember the exact PHP notice, but it occurs when passing a string in directly to `setAdvice`.

Reply:
Ah, gotcha. I just reviewed the use cases for `setAdvice` and there's no need for a reference to be used; I'd recommend getting rid of it (along with the `$msg`).

Reply:
Get rid of `setAdvice`?

Reply:
Sorry, I wasn't clear -- no, just the reference on the second parameter.

Reply:
So it isn't a warning but a fatal error. Just to make sure we understand each other, here's the change I made:

```diff
diff --git a/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php b/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
index 76e3a2ad5..30b3c47e8 100644
--- a/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
+++ b/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php
@@ -49,7 +49,7 @@ class RepresentationUploadAccessPolicy extends DataObjectRequiredPolicy {
 		}
 
 		if (empty($this->_representationId))  {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.representationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
@@ -57,31 +57,31 @@ class RepresentationUploadAccessPolicy extends DataObjectRequiredPolicy {
 		$representation = $representationDao->getById($this->_representationId);
 
 		if (!$representation) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.representationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.representationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
 		$submission = $this->getAuthorizedContextObject(ASSOC_TYPE_SUBMISSION);
 		if (!$submission) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.invalidSubmission');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.invalidSubmission');
 			return AUTHORIZATION_DENY;
 		}
 
 		$publication = Services::get('publication')->get($representation->getData('publicationId'));
 		if (!$publication) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'galley.publicationNotFound');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'galley.publicationNotFound');
 			return AUTHORIZATION_DENY;
 		}
 
 		// Publication and submission must match
 		if ($publication->getData('submissionId') !== $submission->getId()) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'user.authorization.invalidPublication');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'user.authorization.invalidPublication');
 			return AUTHORIZATION_DENY;
 		}
 
 		// Representations can not be modified on published publications
 		if ($publication->getData('status') === STATUS_PUBLISHED) {
-			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, $msg = 'galley.editPublishedDisabled');
+			$this->setAdvice(AUTHORIZATION_ADVICE_DENY_MESSAGE, 'galley.editPublishedDisabled');
 			return AUTHORIZATION_DENY;
 		}
 ```

And here's the fatal error it produces:

```
[15-Oct-2020 13:42:43 Europe/London] PHP Fatal error:  Uncaught Error: Cannot pass parameter 2 by reference in /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php:60
Stack trace:
#0 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.inc.php(77): RepresentationUploadAccessPolicy->dataObjectEffect()
#1 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/AuthorizationDecisionManager.inc.php(196): DataObjectRequiredPolicy->effect()
#2 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/AuthorizationDecisionManager.inc.php(125): AuthorizationDecisionManager->_decidePolicySet(Object(PolicySet), NULL)
#3 /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/handler/PKPHandler.inc.php(299): AuthorizationDecisionManager->decide()
#4 /home/nate/Projects/pkp/ojs-stable/lib/pkp/controllers/wizard/fileUpload/PKPFileUploadWizardHandler.inc.php(176): PKPHandler->authorize(Object(Request), Array, Array)
#5 /home/nate/Pro in /home/nate/Projects/pkp/ojs-stable/lib/pkp/classes/security/authorization/internal/RepresentationUploadAccessPolicy.inc.php on line 60
```

This is with PHP 7.4.10.

Reply:
You can also remove the reference from the `setAdvice` function: https://github.com/pkp/pkp-lib/blob/master/classes/security/authorization/AuthorizationPolicy.inc.php#L58

Reply:
Oh, yeah, that makes sense. Ok, I've updated all of the PRs.